### PR TITLE
feat: Add Wallchain route for selected pair

### PR DESCRIPTION
**Pull request by chef-jojo**
Fix conflict and add a flag to toggle.

Closes #7255 
Closes #7689

### Summary
🚀🔑📦

This pull request adds support for the Wallchain feature, which allows users to swap tokens with better rates using the Wallchain API. It modifies several components, hooks, and files related to the swap view, the swap call, and the approval modal. It also adds or updates some dependencies, environment variables, and translations.

> _We're swapping tokens with the Wallchain_
> _We're getting bonus routes for more gain_
> _We're using hooks and types to handle the trade_
> _We're facing the doom of the gas fees we paid_

### Walkthrough
*  Add Wallchain feature for finding bonus routes for swaps ([link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-c8e463bbf676c362fbc68386873a43c51bfb1a1bb136822808f18c7f2c9fb591L7-R8), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-14b60f636e1a2b0061da57aaf231cb1ed15a5dc0c592425ed82e58fec95d42d8L60-R63), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-7127c22bf51a95345db7f8b59a664483437c11c0a5559150892ff574d4e739feR1-R13), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-2cf2c8889828d5efc330b74403e8b7b87b05ece4edb269a9f351179f300937d1L6-R15), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-2cf2c8889828d5efc330b74403e8b7b87b05ece4edb269a9f351179f300937d1L19-R26), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-2cf2c8889828d5efc330b74403e8b7b87b05ece4edb269a9f351179f300937d1L34-R50), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-c83e4eec2c74b991702f46350b3eee8c029e14df4a79fae45f480542a806111dR34-R35), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-c83e4eec2c74b991702f46350b3eee8c029e14df4a79fae45f480542a806111dR79-R80), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-c83e4eec2c74b991702f46350b3eee8c029e14df4a79fae45f480542a806111dR110), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-d80ee63cb0033c8371bfe572adb8191b01ba0b78bfe7dd5e566f37ad334ba234R48), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-d80ee63cb0033c8371bfe572adb8191b01ba0b78bfe7dd5e566f37ad334ba234L84-R92), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-d80ee63cb0033c8371bfe572adb8191b01ba0b78bfe7dd5e566f37ad334ba234R102), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-d80ee63cb0033c8371bfe572adb8191b01ba0b78bfe7dd5e566f37ad334ba234L104-R130), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-d80ee63cb0033c8371bfe572adb8191b01ba0b78bfe7dd5e566f37ad334ba234L120-L122), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-d80ee63cb0033c8371bfe572adb8191b01ba0b78bfe7dd5e566f37ad334ba234R165-R169), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-d80ee63cb0033c8371bfe572adb8191b01ba0b78bfe7dd5e566f37ad334ba234L150-R179), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-d80ee63cb0033c8371bfe572adb8191b01ba0b78bfe7dd5e566f37ad334ba234L160-R187), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-d80ee63cb0033c8371bfe572adb8191b01ba0b78bfe7dd5e566f37ad334ba234L331-R360), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-83218ea7d05848674a6cd69042376956fd3dfcb9f694f65f7954867778363739L30-R44), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-83218ea7d05848674a6cd69042376956fd3dfcb9f694f65f7954867778363739L50-R55), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-83218ea7d05848674a6cd69042376956fd3dfcb9f694f65f7954867778363739L70-R80), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-83218ea7d05848674a6cd69042376956fd3dfcb9f694f65f7954867778363739L113-R148), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-83218ea7d05848674a6cd69042376956fd3dfcb9f694f65f7954867778363739L190-R215), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-321db73a9e8403e735d93361e2ff4668975ba4fd47b8287df235f6a37788086bR12), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-321db73a9e8403e735d93361e2ff4668975ba4fd47b8287df235f6a37788086bL18-R26), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-321db73a9e8403e735d93361e2ff4668975ba4fd47b8287df235f6a37788086bR33), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-321db73a9e8403e735d93361e2ff4668975ba4fd47b8287df235f6a37788086bL36-R55), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-321db73a9e8403e735d93361e2ff4668975ba4fd47b8287df235f6a37788086bR63-R64), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-321db73a9e8403e735d93361e2ff4668975ba4fd47b8287df235f6a37788086bL61-R96), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-bfe39efb83c5d5d051644d7ebcf06f35473f3a7d27706eae1706fd5e95acac8cR1-R286), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-eb2ab983e2cefc22516cb7814c86346f4b2bd2b971980952868173095d48f459R2703-R2706), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-55650c5c9b98fc543ae9c2251f78f3381d1875b8787150af405987b40683d5ecL9-R12), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-55650c5c9b98fc543ae9c2251f78f3381d1875b8787150af405987b40683d5ecL31-R37), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbR631-R633), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbR640-R642), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbR10831-R10836), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbR11318-R11327), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbR12022-R12037), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbR13840-R13843), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL14340), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL15419), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL18081-R18121), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL18442-R18482), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL18658), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL21742), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL22985-R23023), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbR23548-R23563), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL24030), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbR25029-R25031), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL26935), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL27089), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL27121), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL27163))
  * Add a new environment variable for the Wallchain BSC key in `apps/web/.env.example` ([link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-c8e463bbf676c362fbc68386873a43c51bfb1a1bb136822808f18c7f2c9fb591L7-R8))
  * Add a new dependency for the `@wallchain/sdk` package in `apps/web/package.json` and `pnpm-lock.yaml` ([link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-14b60f636e1a2b0061da57aaf231cb1ed15a5dc0c592425ed82e58fec95d42d8L60-R63), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbR631-R633), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbR12022-R12037))
  * Add a new file for the Wallchain configuration in `apps/web/src/config/wallchain.ts` ([link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-7127c22bf51a95345db7f8b59a664483437c11c0a5559150892ff574d4e739feR1-R13))
  * Add a new prop for the `RoutesBreakdown` component to indicate whether the trade is a market maker trade ([link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-2cf2c8889828d5efc330b74403e8b7b87b05ece4edb269a9f351179f300937d1L19-R26))
  * Modify the text for the route label and the question helper in `apps/web/src/views/Swap/V3Swap/components/RoutesBreakdown.tsx` based on the trade type ([link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-2cf2c8889828d5efc330b74403e8b7b87b05ece4edb269a9f351179f300937d1L34-R50))
  * Add new hooks for checking the Wallchain status and using the Wallchain API in `apps/web/src/views/Swap/V3Swap/hooks/useWallchain.ts` ([link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-bfe39efb83c5d5d051644d7ebcf06f35473f3a7d27706eae1706fd5e95acac8cR1-R286))
  * Add new state variables for the Wallchain status and the bonus route flag in `apps/web/src/views/Swap/V3Swap/containers/ConfirmSwapModal.tsx` ([link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-c83e4eec2c74b991702f46350b3eee8c029e14df4a79fae45f480542a806111dR79-R80))
  * Add a new prop for the `ConfirmSwapModal` component to indicate whether the trade is a bonus route trade ([link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-c83e4eec2c74b991702f46350b3eee8c029e14df4a79fae45f480542a806111dR110))
  * Add a new hook for using the Wallchain API and getting the approval address and the master input in `apps/web/src/views/Swap/V3Swap/containers/SwapCommitButton.tsx` ([link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-d80ee63cb0033c8371bfe572adb8191b01ba0b78bfe7dd5e566f37ad334ba234R48))
  * Add new variables for the deadline, the approval address, the master input, the approval submitted, and the revert reason in `apps/web/src/views/Swap/V3Swap/containers/SwapCommitButton.tsx` ([link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-d80ee63cb0033c8371bfe572adb8191b01ba0b78bfe7dd5e566f37ad334ba234L84-R92), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-d80ee63cb0033c8371bfe572adb8191b01ba0b78bfe7dd5e566f37ad334ba234L104-R130), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-d80ee63cb0033c8371bfe572adb8191b01ba0b78bfe7dd5e566f37ad334ba234L120-L122))
  * Modify the router address and the swap callback based on the trade type in `apps/web/src/views/Swap/V3Swap/containers/SwapCommitButton.tsx` ([link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-d80ee63cb0033c8371bfe572adb8191b01ba0b78bfe7dd5e566f37ad334ba234R102), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-d80ee63cb0033c8371bfe572adb8191b01ba0b78bfe7dd5e566f37ad334ba234L120-L122))
  * Add new logic for handling the swap callback error and success based on the trade type in `apps/web/src/views/Swap/V3Swap/containers/SwapCommitButton.tsx` ([link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-d80ee63cb0033c8371bfe572adb8191b01ba0b78bfe7dd5e566f37ad334ba234R165-R169), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-d80ee63cb0033c8371bfe572adb8191b01ba0b78bfe7dd5e566f37ad334ba234L150-R179))
  * Add a new dependency for the handle confirm dismiss function in `apps/web/src/views/Swap/V3Swap/containers/SwapCommitButton.tsx` ([link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-d80ee63cb0033c8371bfe572adb8191b01ba0b78bfe7dd5e566f37ad334ba234L160-R187))
  * Add a new condition for disabling the swap button based on the Wallchain status in `apps/web/src/views/Swap/V3Swap/containers/SwapCommitButton.tsx` ([link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-d80ee63cb0033c8371bfe572adb8191b01ba0b78bfe7dd5e566f37ad334ba234L331-R360))
  * Add the `WallchainSwapCall` type as a possible value for the swap calls in `apps/web/src/views/Swap/V3Swap/hooks/useSendSwapTransaction.ts` ([link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-83218ea7d05848674a6cd69042376956fd3dfcb9f694f65f7954867778363739L30-R44), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-83218ea7d05848674a6cd69042376956fd3dfcb9f694f65f7954867778363739L50-R55))
  * Add a new condition for mapping the swap calls based on the call type in `apps/web/src/views/Swap/V3Swap/hooks/useSendSwapTransaction.ts` ([link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-83218ea7d05848674a6cd69042376956fd3dfcb9f694f65f7954867778363739L70-R80))
  * Modify the best call option and the swap callback based on the call type in `apps/web/src/views/Swap/V3Swap/hooks/useSendSwapTransaction.ts` ([link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-83218ea7d05848674a6cd69042376956fd3dfcb9f694f65f7954867778363739L113-R148), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-83218ea7d05848674a6cd69042376956fd3dfcb9f694f65f7954867778363739L190-R215))
  * Add a new type alias for the master input and a new property for the revert reason in `apps/web/src/views/Swap/V3Swap/hooks/useSwapCallback.ts` ([link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-321db73a9e8403e735d93361e2ff4668975ba4fd47b8287df235f6a37788086bR12), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-321db73a9e8403e735d93361e2ff4668975ba4fd47b8287df235f6a37788086bL18-R26), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-321db73a9e8403e735d93361e2ff4668975ba4fd47b8287df235f6a37788086bR33))
  * Add new properties for the on wallchain drop function and the master input in `apps/web/src/views/Swap/V3Swap/hooks/useSwapCallback.ts` ([link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-321db73a9e8403e735d93361e2ff4668975ba4fd47b8287df235f6a37788086bL36-R55))
  * Add new arguments for the use swap callback hook in `apps/web/src/views/Swap/V3Swap/hooks/useSwapCallback.ts` ([link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-321db73a9e8403e735d93361e2ff4668975ba4fd47b8287df235f6a37788086bR63-R64))
  * Modify the swap calls based on the trade type in `apps/web/src/views/Swap/V3Swap/hooks/useSwapCallback.ts` ([link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-321db73a9e8403e735d93361e2ff4668975ba4fd47b8287df235f6a37788086bL61-R96))
  * Add new translations for the MM Route and the Bonus Route in `packages/localization/src/config/translations.json` ([link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-eb2ab983e2cefc22516cb7814c86346f4b2bd2b971980952868173095d48f459R2703-R2706))
  * Add a new prop for the `ApproveModalContent` component to indicate whether the trade is a bonus route trade ([link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-55650c5c9b98fc543ae9c2251f78f3381d1875b8787150af405987b40683d5ecL9-R12))
  * Modify the text for the approve modal content based on the trade type in `packages/uikit/src/widgets/Swap/ApproveModalContent.tsx` ([link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-55650c5c9b98fc543ae9c2251f78f3381d1875b8787150af405987b40683d5ecL31-R37))
  * Add new dependencies for the `bottleneck`, `@types/randombytes`, and `@uniswap/permit2-sdk` packages in `pnpm-lock.yaml` ([link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbR640-R642), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbR10831-R10836), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbR11318-R11327), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbR13840-R13843))
  * Remove the dev flag from some packages in `pnpm-lock.yaml` ([link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL14340), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL15419), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL18658), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL21742), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL26935), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL27089), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL27121), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL27163))
  * Update the `tiny-invariant` dependency in `pnpm-lock.yaml` ([link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL18081-R18121), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL22985-R23023), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbR25029-R25031))
  * Add the hasBin flag to the `is-glob` package in `pnpm-lock.yaml` ([link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL18442-R18482))
  * Add a new dependency for the `rollup-plugin-visualizer` package in `pnpm-lock.yaml` ([link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbR23548-R23563))
  * Remove the dev flag from the `tiny-invariant` package in `pnpm-lock.yaml` ([link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL24030))
* Modify the text for the route label in `apps/web/src/views/Swap/components/AdvancedSwapDetails.tsx` to use the MM Route translation ([link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-65730801b841c0fd17e97ac2609066b79fd2875585af9babde2b88f1481348e0L198-R203))
* Add optional chaining operators for the input and output currencies in `apps/web/src/views/Swap/components/AdvancedSwapDetails.tsx` to avoid potential errors ([link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-65730801b841c0fd17e97ac2609066b79fd2875585af9babde2b88f1481348e0L57-R58), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-65730801b841c0fd17e97ac2609066b79fd2875585af9babde2b88f1481348e0L141-R141), [link](https://github.com/pancakeswap/pancake-frontend/pull/7931/files?diff=unified&w=0#diff-65730801b841c0fd17e97ac2609066b79fd2875585af9babde2b88f1481348e0L226-R231))

### CODE REVIEW

**Comment by chef-jojo** [comment]
@SocketSecurity ignore rollup-plugin-visualizer@5.9.2

File: apps/web/src/config/wallchain.ts [hunk 1]
```diff
@@ -0,0 +1,13 @@
+import { bscTokens } from '@pancakeswap/tokens'
+import { ChainId, ERC20Token } from '@pancakeswap/sdk'
```
Comment (ChefJerry) [inline apps/web/src/config/wallchain.ts:2]:
```suggestion
import { ERC20Token } from '@pancakeswap/sdk'
import { ChainId } from '@pancakeswap/chains'
```

File: apps/web/src/views/Swap/V3Swap/containers/SwapCommitButton.tsx [hunk 2]
```diff
@@ -81,7 +82,14 @@ export const SwapCommitButton = memo(function SwapCommitButton({
   const showWrap = wrapType !== WrapType.NOT_APPLICABLE
   const [isRoutingSettingChange, resetRoutingSetting] = useRoutingSettingChanged()
   const slippageAdjustedAmounts = useSlippageAdjustedAmounts(trade)
-  const routerAddress = SMART_ROUTER_ADDRESSES[trade?.inputAmount?.currency?.chainId]
+
+  const deadline = useTransactionDeadline()
+  const [statusWallchain, approvalAddressForWallchain, wallchainMasterInput] = useWallchainApi(trade, deadline)
+  const [wallchainSecondaryStatus, setWallchainSecondaryStatus] = useState('not-found')
```
Comment (ChefJerry) [inline apps/web/src/views/Swap/V3Swap/containers/SwapCommitButton.tsx:88]:
```suggestion
  const [wallchainSecondaryStatus, setWallchainSecondaryStatus] = useState<'found'|'not-found'>('not-found')
```

why not just use Boolean

File: apps/web/src/views/Swap/V3Swap/containers/SwapCommitButton.tsx [hunk 3]
```diff
@@ -142,22 +162,29 @@ export const SwapCommitButton = memo(function SwapCommitButton({
...
       .then((res) => {
+        setWallchainSecondaryStatus('not-found')
         setSwapState({ attemptingTxn: false, tradeToConfirm, swapErrorMessage: undefined, txHash: res.hash })
       })
       .catch((error) => {
+        setWallchainSecondaryStatus('not-found')
         setSwapState({
           attemptingTxn: false,
           tradeToConfirm,
           swapErrorMessage: typeof error === 'string' ? error : error.message,
```
Comment (ChefJerry) [inline apps/web/src/views/Swap/V3Swap/containers/SwapCommitButton.tsx:183]:
```suggestion
          swapErrorMessage: typeof error === 'string' ? error : error?.message,
```

error not means Error type in catch clause, always use `?.` to retrive message

File: apps/web/src/views/Swap/V3Swap/hooks/useWallchain.ts [hunk 4]
```diff
@@ -0,0 +1,287 @@
+import { useState, useEffect, useMemo } from 'react'
+import type WallchainSDK from '@wallchain/sdk'
+import type { TMEVFoundResponse } from '@wallchain/sdk'
+import { TOptions } from '@wallchain/sdk'
+import { Token, TradeType, Currency, ChainId } from '@pancakeswap/sdk'
```
Comment (ChefJerry) [inline apps/web/src/views/Swap/V3Swap/hooks/useWallchain.ts:5]:
```suggestion
import { Token, TradeType, Currency } from '@pancakeswap/sdk'
import { ChainId } from '@pancakeswap/chains'
```

File: apps/web/src/views/Swap/V3Swap/hooks/useWallchain.ts [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+          setSwapCalls([{ getCall: callback }])
+        }
+      } else {
+        const callback = async () => {
+          try {
+            const spender = (await sdk.getSpender()) as `0x${string}`
+            let witness: false | Awaited<ReturnType<typeof sdk.signPermit>> = false
+
+            if (needPermit) {
+              witness = await sdk.signPermit(srcToken as `0x${string}`, account, spender, amountIn)
```
Comment (ChefJerry) [inline apps/web/src/views/Swap/V3Swap/hooks/useWallchain.ts:241]: no need add as anymore, srcToken already have type `false | `0x${string}``

File: apps/web/src/views/Swap/V3Swap/hooks/useWallchain.ts [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+
+            const data = await sdk.createNewTransaction(
+              previousSwapCalls[0].address,
+              previousSwapCalls[0].address,
+              false,
+              previousSwapCalls[0].calldata,
+              amountIn,
+              previousSwapCalls[0].value,
+              srcToken as `0x${string}`,
+              dstToken as `0x${string}`,
```
Comment (ChefJerry) [inline apps/web/src/views/Swap/V3Swap/hooks/useWallchain.ts:252]: also can remove `as`

File: apps/web/package.json [hunk 7]
```diff
@@ -57,8 +57,10 @@
     "@snapshot-labs/snapshot.js": "^0.4.40",
     "@tanstack/react-query": "^4.28.0",
     "@vanilla-extract/next-plugin": "^2.3.0",
+    "@wallchain/sdk": "^0.6.8",
     "ajv": "^6.12.3",
     "bignumber.js": "^9.0.0",
+    "bottleneck": "^2.19.5",
```
Comment (ChefJerry) [inline apps/web/package.json:63]: last publish 4 years ago 🥲

File: packages/localization/src/config/translations.json [hunk 8]
```diff
@@ -2704,6 +2704,10 @@
   "Mercury Mysteries": "Mercury Mysteries",
   "Venus Protectors": "Venus Protectors",
   "Uranus Unity Rain": "Uranus Unity Rain",
+  "MM Route": "MM Route",
```
Comment (ChefJerry) [inline packages/localization/src/config/translations.json:2707]: type check of i18n key now is failed, hard to review if missed key

Comment (chef-jojo) [reply packages/localization/src/config/translations.json:2707]: can you explain more?

Comment (ChefJerry) [reply packages/localization/src/config/translations.json:2707]: `t('xxx')` if xxx is not one key of the json

Comment (chef-jojo) [reply packages/localization/src/config/translations.json:2707]: It's not meant to be strict type safe by using (string || {}) for selection complete only

We have our own test cases for it

https://github.com/pancakeswap/pancake-frontend/blob/develop/apps/web/src/__tests__/translations.test.ts

**Pull request merged**